ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/MerkleTree.java

```diff
@@ -38,45 +38,110 @@ protected MerkleTree(int treeDepth) {
     zeroHashes = generateZeroHashes(treeDepth);
   }
 
-  abstract void add(Bytes32 leaf);
+  public abstract void add(Bytes32 leaf);
 
-  protected abstract int getNumberOfLeaves();
+  public abstract int getNumberOfLeaves();
 
   protected static List<Bytes32> generateZeroHashes(int height) {
     List<Bytes32> zeroHashes = new ArrayList<>();
-    for (int i = 0; i < height; i++) {
-      zeroHashes.add(Bytes32.ZERO);
-    }
-    for (int i = 0; i < height - 1; i++) {
-      zeroHashes.set(i + 1, Hash.sha2_256(Bytes.concatenate(zeroHashes.get(i), zeroHashes.get(i))));
+    zeroHashes.add(Bytes32.ZERO);
+    for (int i = 1; i < height; i++) {
+      zeroHashes.add(
+          i, Hash.sha2_256(Bytes.concatenate(zeroHashes.get(i - 1), zeroHashes.get(i - 1))));
     }
     return zeroHashes;
   }
 
-  public SSZVector<Bytes32> getProofTreeByValue(Bytes32 value) {
+  public SSZVector<Bytes32> getProof(Bytes32 value) {
     int index = tree.get(0).indexOf(value);
-    return getProofTreeByIndex(index);
+    return getProof(index);
   }
 
-  public SSZVector<Bytes32> getProofTreeByIndex(int index) {
+  public SSZVector<Bytes32> getProof(int index) {
     List<Bytes32> proof = new ArrayList<>();
     for (int i = 0; i < treeDepth; i++) {
       index = index % 2 == 1 ? index - 1 : index + 1;
-      if (index < tree.get(i).size()) proof.add(tree.get(i).get(index));
-      else proof.add(zeroHashes.get(i));
+      if (index < tree.get(i).size()) {
+        proof.add(tree.get(i).get(index));
+      } else {
+        proof.add(zeroHashes.get(i));
+      }
       index /= 2;
     }
     proof.add(calcMixInValue());
     return new SSZVector<>(proof, Bytes32.class);
   }
 
-  private Bytes32 calcMixInValue() {
+  private Bytes32 calcViewBoundaryRoot(int depth, int viewLimit) {
+    if (depth == 0) {
+      return zeroHashes.get(0);
+    }
+    depth -= 1;
+    Bytes32 deeperRoot = calcViewBoundaryRoot(depth, viewLimit);
+    if ((viewLimit & (1 << depth)) != 0) {
+      return Hash.sha2_256(Bytes.concatenate(tree.get(depth).get(viewLimit >> depth), deeperRoot));
+    } else {
+      return Hash.sha2_256(Bytes.concatenate(deeperRoot, zeroHashes.get(depth)));
+    }
+  }
+
+  public SSZVector<Bytes32> getProofWithViewBoundary(Bytes32 value, int viewLimit) {
+    return getProofWithViewBoundary(tree.get(0).indexOf(value), viewLimit);
+  }
+
+  public SSZVector<Bytes32> getProofWithViewBoundary(int index, int viewLimit) {
+    checkArgument(index < viewLimit, "MerkleTree: Index must be within view limit");
+
+    List<Bytes32> proof = new ArrayList<>();
+    for (int i = 0; i < treeDepth; i++) {
+      // Get index of sibling node
+      index = index % 2 == 1 ? index - 1 : index + 1;
+
+      // Check how much of the tree at this level is strictly within the view limit.
+      int limit = viewLimit >> i;
+
+      checkArgument(
+          limit <= tree.get(i).size(), "MerkleTree: Tree is too small for given limit at height");
```

Comment:
Is this equivalent to checking that viewLimit is <= # of leaf nodes?  Could we just have a single check at the top for this?

Reply:
I am not sure if the two are equivalent